### PR TITLE
Kubernetes decorator

### LINKED ISSUE #19135: PythonKubernetesOperator and kubernetes taskflow decorator
### Description

After the implementation of the Docker Taskflow Operator in 2.2, having a simple function on Kubernetes should be quite simple to accomplish.

One major difference I see is how to get the Code into the Pod.
My prefered way here would be to add an init container which takes care of this - much similar on how we use a Sidecar to extract XCom.

I would also prefer to add more defaults than the KubernetesPodOperator does by default to keep the call as simple and straightforward as possible. We could for example default the namespace to the namespace Airflow is running in if it is a Kubernetes Deployment. Also, we could generate a reasonable pod name from DAG-ID and Task ID.

### Use case/motivation

Beeing able to run a task as simple as:

```python
@task.kubernetes(image='my_image:1.1.0')
def sum(a, b):
    from my_package_in_version_1 import complex_sum
    return complex_sum(a,b)
```
would be awesome!
It would cleanly separate environments and resources between tasks just as the Docker or Virtualenv Operators do - and unlinke the `@task.python` decorator. We could specify the resources and environment of each task freely by specifying a `resources` Argument.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by subkanthi**
Decorator to execute functions in k8s using KubernetesPodOperator

closes: #19135
```
@task.kubernetes(image='python:3.8-slim-buster', name='k8s_test', namespace='default')
def k8s_decorator_func():
       print("decorator func")
```

If no parameters passed, defaults are set in kubernetes.py
```
        # Image, name and namespace are all required.
        if not 'image' in kwargs:
            kwargs['image'] = 'python:3.8-slim-buster'

        if not 'name' in kwargs:
            kwargs['name'] = f'k8s_airflow_pod_{uuid.uuid4().hex}'

        if not 'namespace' in kwargs:
            kwargs['namespace'] = 'default'
```

This version sets the function as an environment variable - __PYTHON_SCRIPT__ and in the command , the logic is to retrieve the environment variable and write to '/tmp/script.py' to execute the function(very similar to Docker)

1) Initially proceeded with trying to mount a local file using the hostPath, the problem is that requires Airflow also to run in the same node. For example running airflow locally and connecting to the kind cluster would need a mount.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                jinja_context=jinja_context,
+                filename=script_filename,
+                template_file='python_kubernetes_script.jinja2',
+            )
+
+            self.env_vars.append(
+                k8s.V1EnvVar(name="__PYTHON_SCRIPT", value=_read_file_contents(script_filename))
+            )
+
+            self.cmds.append("bash")
```
Comment (subkanthi) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py:109]: Thoughts here, The user can still pass cmds and arguments, 2 options, either we limit the user not to pass command and arguments or check here if its passed and not append.

File: airflow/decorators/__init__.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 from airflow.decorators.python_virtualenv import virtualenv_task
 from airflow.decorators.task_group import task_group
 from airflow.models.dag import dag
+from airflow.providers.cncf.kubernetes.decorators.kubernetes import kubernetes_task
```
Comment (uranusjr) [inline airflow/decorators/__init__.py]: This can’t be done like this here because the user may not have the kubernetes provider installed. You need to use the provider mechanism to declare the decorator. See [development documentation](https://airflow.apache.org/docs/apache-airflow/stable/howto/create-custom-decorator.html) for a more detailed explaination, and how the `@task.docker` decorator is implemented as an example.

File: airflow/decorators/__init__.pyi [hunk 3]
```diff
@@ -126,7 +126,25 @@ class TaskDecoratorCollection:
             such as transmission a large amount of XCom to TaskAPI.
         """
     @overload
-    def virtualenv(self, python_callable: Function) -> Function: ...
+    def virtualenv(self, python_callable: F) -> F: ...
+    @overload
+    def kubernetes(
+        python_callable: Optional[Callable] = None, multiple_outputs: Optional[bool] = None, **kwargs
+    ) -> TaskDecorator:
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: This is wrong; again, see the documentation and the Docker decorator.

File: airflow/utils/python_kubernetes_script.jinja2 [hunk 4]
```diff
@@ -0,0 +1,44 @@
+{#
```
Comment (uranusjr) [inline airflow/utils/python_kubernetes_script.jinja2:1]: This entire file should be placed in the kubernetes provider.

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Comment by uranusjr** [comment]
```
airflow/example_dags/example_kubernetes_decorator.py:32:6: F811 redefinition of unused 'dag' from line 25
airflow/providers/cncf/kubernetes/decorators/kubernetes.py:81:12: E713 test for membership should be 'not in'
airflow/providers/cncf/kubernetes/decorators/kubernetes.py:84:12: E713 test for membership should be 'not in'
airflow/providers/cncf/kubernetes/decorators/kubernetes.py:87:12: E713 test for membership should be 'not in'
airflow/providers/cncf/kubernetes/decorators/kubernetes.py:90:9: F841 local variable 'kwargs_to_upstream' is assigned to but never used
```

File: airflow/utils/python_kubernetes_script.jinja2 (hunk 4 above)
Comment (subkanthi) [reply airflow/utils/python_kubernetes_script.jinja2:1]: currently this template is loaded using the
`write_python_script` function in python_virtualenv.py, it shared by virtualenv and docker decorators. Thats why the jinja template needs to be in the same folder. 
Ideally we should move this function out of python_virtualenv.

File: airflow/decorators/__init__.pyi (hunk 3 above)
Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Moved, thanks

File: airflow/decorators/__init__.py (hunk 2 above)
Comment (subkanthi) [reply airflow/decorators/__init__.py]: Removed.

File: airflow/decorators/__init__.pyi [hunk 5]
```diff
@@ -126,7 +126,22 @@ class TaskDecoratorCollection:
             such as transmission a large amount of XCom to TaskAPI.
         """
     @overload
-    def virtualenv(self, python_callable: Function) -> Function: ...
+    def virtualenv(self, python_callable: F) -> F: ...
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: Why is this line changed? There is no `F`.

File: airflow/decorators/__init__.pyi [hunk 6]
```diff
@@ -126,7 +126,22 @@ class TaskDecoratorCollection:
...
+        Also accepts any argument that KubernetesPodOperator will via ``kwargs``.
+
+        :param python_callable: Function to decorate
+        :type python_callable: Optional[Callable]
+        :param multiple_outputs: if set, function return value will be
+            unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+            with index as key. Dict will unroll to xcom values with keys as XCom keys.
+            Defaults to False.
+        :type multiple_outputs: bool
+        """
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: This should explicitly spell out most (or all) of the arguments on `KubernetesPodOperator`, like `docker` below.

Also, can you move this to after `docker`? This better separates built-in and provider-installed decorators.

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/decorators/__init__.pyi (hunk 5 above)
Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Reverted change.

File: airflow/decorators/__init__.pyi (hunk 6 above)
Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Added, thanks.

**Comment by subkanthi** [comment]
> ```
> airflow/example_dags/example_kubernetes_decorator.py:32:6: F811 redefinition of unused 'dag' from line 25
> airflow/providers/cncf/kubernetes/decorators/kubernetes.py:81:12: E713 test for membership should be 'not in'
> airflow/providers/cncf/kubernetes/decorators/kubernetes.py:84:12: E713 test for membership should be 'not in'
> airflow/providers/cncf/kubernetes/decorators/kubernetes.py:87:12: E713 test for membership should be 'not in'
> airflow/providers/cncf/kubernetes/decorators/kubernetes.py:90:9: F841 local variable 'kwargs_to_upstream' is assigned to but never used
> ```

Fixed these errors.

**Comment by subkanthi** [comment]
Cant reproduce the mypy errors locally,  thoughts @uranusjr 
```Run mypy.................................................................................

The CI image for Python 3.7 might be outdated

Please run this command at earliest convenience: 

./breeze build-image --python 3.7



Passed```

**Comment by potiuk** [comment]
Did you dollow the advice @subkanthi ?

```
./breeze build-image --python 3.7
```

**Comment by subkanthi** [comment]
> Did you dollow the advice @subkanthi ?
> 
> ```
> ./breeze build-image --python 3.7
> ```

Not really, sorry will try again.

**Comment by subkanthi** [comment]
> > Did you dollow the advice @subkanthi ?
> > ```
> > ./breeze build-image --python 3.7
> > ```
> 
> Not really, sorry will try again.

unfortunately that didnt help, will try to troubleshoot the problem.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py (hunk 1 above)
Comment (uranusjr) [reply airflow/providers/cncf/kubernetes/decorators/kubernetes.py:109]: I’d say let’s not allow passing the command for now, and loosen the restriction if anyone asks for it.

**Comment by potiuk** [comment]
@uranusjr  - looks ready for re-review :)

File: airflow/decorators/__init__.pyi [hunk 7]
```diff
@@ -243,5 +243,123 @@ class TaskDecoratorCollection:
...
         """
         # [END decorator_signature]
+    def kubernetes(
+        self,
+        python_callable: Optional[Callable] = None,
+        multiple_outputs: Optional[bool] = None,
+        namespace: Optional[str] = None,
+        image: Optional[str] = None,
+        name: Optional[str] = None,
+        random_name_suffix: Optional[bool] = True,
```
Comment (josh-fell) [inline airflow/decorators/__init__.pyi]:
```suggestion
        random_name_suffix: bool = True,
```

File: airflow/decorators/__init__.pyi [hunk 8]
```diff
@@ -243,5 +243,123 @@ class TaskDecoratorCollection:
...
+
+        Also accepts any argument that KubernetesPodOperator will via ``kwargs``.
+
+        :param python_callable: Function to decorate
+        :type python_callable: Optional[Callable]
+        :param multiple_outputs: if set, function return value will be
+            unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+            with index as key. Dict will unroll to xcom values with keys as XCom keys.
+            Defaults to False.
+        :type multiple_outputs: bool
```
Comment (josh-fell) [inline airflow/decorators/__init__.pyi]:
```suggestion
        :param python_callable: Function to decorate
        :param multiple_outputs: if set, function return value will be
            unrolled to multiple XCom values. Dict will unroll to XCom values with keys as XCom keys.
            Defaults to False.
```
`:type:` directives are no longer required (see #20951) and a pre-commit check was added which should eliminate these as well. Also `multiple_outputs` does not handle lists and tuples.

File: airflow/example_dags/example_kubernetes_decorator.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+from datetime import datetime
+
+from airflow import DAG
+from airflow.decorators import task
+
+with DAG(
+    dag_id='example_kubernetes_decorator',
+    schedule_interval=None,
+    start_date=datetime(2021, 1, 1),
+    tags=['example'],
```
Comment (josh-fell) [inline airflow/example_dags/example_kubernetes_decorator.py:31]:
```suggestion
    tags=['example'],
    catchup=False,
```
Example DAGs should have `catchup=False` just to be proactive with any unforeseen headaches users _might_ have when modifying from this example.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+) -> TaskDecorator:
+    """
+    Kubernetes operator decorator. Wraps a function to be executed in K8s using KubernetesPodOperator.
+    Also accepts any argument that DockerOperator will via ``kwargs``. Can be reused in a single DAG.
+
+    :param python_callable: Function to decorate
+    :param multiple_outputs: if set, function return value will be
+        unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+        with index as key. Dict will unroll to xcom values with keys as XCom keys.
+        Defaults to False.
```
Comment (josh-fell) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py]:
```suggestion
    :param multiple_outputs: if set, function return value will be
        unrolled to multiple XCom values. Dict will unroll to xcom values with keys as XCom keys.
        Defaults to False.
```
Same comment here on types handled with `multiple_outputs`.

**Review by josh-fell — CHANGES_REQUESTED** [review]

**Comment by subkanthi** [comment]
Thanks @josh-fell , updated.

File: airflow/decorators/__init__.pyi [hunk 11]
```diff
@@ -251,5 +251,120 @@ class TaskDecoratorCollection:
         :param cap_add: Include container capabilities
         """
         # [END decorator_signature]
+    def kubernetes(
+        self,
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi:242]:
```suggestion
    def kubernetes(
        self,
        *,
```

Arguments on `KubernetesOperator` are keyword-only, so these need to be as well.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+    def __init__(
+        self,
+        **kwargs,
+    ) -> None:
+        self.pickling_library = pickle
+
+        # Image, name and namespace are all required.
+        if 'image' not in kwargs:
+            kwargs['image'] = 'python:3.8-slim-buster'
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py]: Why 3.8 and why slim? Maybe we should just make this a required argument instead of trying to guess what’s best for the user.

Comment (subkanthi) [reply airflow/providers/cncf/kubernetes/decorators/kubernetes.py]: @uranusjr Added exception if image is not passed.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    template_fields: Sequence[str] = ('op_args', 'op_kwargs')
+
+    # since we won't mutate the arguments, we should just do the shallow copy
+    # there are some cases we can't deepcopy the objects (e.g protobuf).
+    shallow_copy_attrs: Sequence[str] = ('python_callable',)
+
+    def __init__(
+        self,
+        **kwargs,
+    ) -> None:
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py:73]:
```python
def __init__(self, image: str, **kwargs):
```

Also mention `image` is required in the docstring.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+        self.pickling_library = pickle
+
+        # Image, name and namespace are all required.
+        if 'image' not in kwargs:
+            raise AirflowException("Image required for kubernetes decorator")
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py]: And remove this.

File: airflow/decorators/__init__.pyi [hunk 15]
```diff
@@ -245,5 +245,121 @@ class TaskDecoratorCollection:
...
+        *,
+        python_callable: Optional[Callable] = None,
+        multiple_outputs: Optional[bool] = None,
+        namespace: Optional[str] = None,
+        image: Optional[str] = None,
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]:
```suggestion
        image: str,
```

File: airflow/decorators/__init__.pyi [hunk 16]
```diff
@@ -245,5 +245,121 @@ class TaskDecoratorCollection:
         :param cap_add: Include container capabilities
         """
         # [END decorator_signature]
+    def kubernetes(
+        self,
+        *,
+        python_callable: Optional[Callable] = None,
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]: Remove this since the argument-less form is not possible. See the `docker` decorator for an example.

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/decorators/__init__.pyi (hunk 16 above)
Comment (subkanthi) [reply airflow/decorators/__init__.pyi]: Removed, thanks.

File: airflow/decorators/__init__.pyi [hunk 17]
```diff
@@ -188,8 +188,7 @@ class TaskDecoratorCollection:
             Dict will unroll to XCom values with keys as XCom keys. Defaults to False.
         :param use_dill: Whether to use dill or pickle for serialization
         :param python_command: Python command for executing functions, Default: python3
-        :param image: Docker image from which to create the container.
-            If image tag is omitted, "latest" will be used.
+        :param image: Docker image from which to create the container.(Required)
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi]:
```suggestion
        :param image: Docker image from which to create the container. (Required)
```

**Review by uranusjr — CHANGES_REQUESTED** [review]
Also need to add `image` to the argument list of `_KubernetesDecoratedOperator`

File: airflow/utils/python_kubernetes_script.jinja2 (hunk 4 above)
Comment (ashb) [reply airflow/utils/python_kubernetes_script.jinja2:1]: This _has_ to be in the kube provider folder, else it won't get released with the provider, and it won't work.

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py (hunk 1 above)
Comment (Asciotti) [reply airflow/providers/cncf/kubernetes/decorators/kubernetes.py:109]: If I am reading this correctly, we are setting the `KubernetesPod`'s `cmd` to `bash` then some `args` to the remaining pieces. I believe this hit us, we had docker images with `ENTRYPOINTS` and I believe the `cmd` directive here will override those entirely. So by forcing a `bash` here we are killing any `ENTRYPOINTS`. We frequently have `dumb init --` ones, or `poetry run` ones, to make sure we are running our commands safely, and in an isolated manner.

Did I understand this correctly?

**Comment by Asciotti** [comment]
Bump? :)

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py (hunk 1 above)
Comment (Asciotti) [reply airflow/providers/cncf/kubernetes/decorators/kubernetes.py:109]: Nevermind, `append` :)

File: airflow/utils/python_kubernetes_script.jinja2 (hunk 4 above)
Comment (kaxil) [reply airflow/utils/python_kubernetes_script.jinja2:1]: Can you move this @subkanthi to the cncf kube providers please?

**Review by kaxil — COMMENTED** [review]
Please also rebase your PR?

File: airflow/utils/python_kubernetes_script.jinja2 (hunk 4 above)
Comment (kaxil) [reply airflow/utils/python_kubernetes_script.jinja2:1]: This has still not been resolved

File: airflow/decorators/__init__.pyi [hunk 18]
```diff
@@ -239,5 +238,119 @@ class TaskDecoratorCollection:
...
+        full_pod_spec: Optional[k8s.V1Pod] = None,
+        init_containers: Optional[List[k8s.V1Container]] = None,
+        log_events_on_failure: bool = False,
+        do_xcom_push: bool = False,
+        pod_template_file: Optional[str] = None,
+        priority_class_name: Optional[str] = None,
+        pod_runtime_info_envs: Optional[List[PodRuntimeInfoEnv]] = None,
+        termination_grace_period: Optional[int] = None,
+        configmaps: Optional[List[str]] = None,
+        **kwargs,
```
Comment (uranusjr) [inline airflow/decorators/__init__.pyi:287]: What can go into `kwargs` here? It seems unnecessary.

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        **kwargs,
+    ) -> None:
+        self.pickling_library = pickle
+
+        # Set defaults for name and namespace.
+        if 'name' not in kwargs:
+            kwargs['name'] = f'k8s_airflow_pod_{uuid.uuid4().hex}'
+
+        if 'namespace' not in kwargs:
+            kwargs['namespace'] = 'default'
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py:81]: Nit, it’s easier to use `setdefault` for these

File: airflow/providers/cncf/kubernetes/decorators/kubernetes.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+            return super().execute(context)
+
+    def _get_python_source(self):
+        raw_source = inspect.getsource(self.python_callable)
+        res = dedent(raw_source)
+        res = remove_task_decorator(res, "@task.kubernetes")
+        return res
+
+
+T = TypeVar("T", bound=Callable)
```
Comment (uranusjr) [inline airflow/providers/cncf/kubernetes/decorators/kubernetes.py:125]: Unused?

**Comment by potiuk** [comment]
Closing, as this has been already implemented and merged in #25663
